syft/pkg/cataloger/deb/cataloger.go

```diff
@@ -56,93 +59,166 @@ func (c *Cataloger) Catalog(resolver source.FileResolver) ([]pkg.Package, error)
 			p.FoundBy = c.Name()
 			p.Locations = []source.Location{dbLocation}
 
-			metadata := p.Metadata.(pkg.DpkgMetadata)
+			// the current entry only has what may have been listed in the status file, however, there are additional
+			// files that are listed in multiple other locations. We should retrieve them all and merge the file lists
+			// together.
+			mergeFileListing(resolver, dbLocation, p)
 
-			md5Reader, md5Location, err := fetchMd5Contents(resolver, dbLocation, p)
-			if err != nil {
-				return nil, fmt.Errorf("unable to find dpkg md5 contents: %w", err)
-			}
+			// fetch additional data from the copyright file to derive the license information
+			addLicenses(resolver, dbLocation, p)
+		}
 
-			if md5Reader != nil {
-				// attach the file list
-				metadata.Files = parseDpkgMD5Info(md5Reader)
-
-				// keep a record of the file where this was discovered
-				if md5Location != nil {
-					p.Locations = append(p.Locations, *md5Location)
-				}
-			} else {
-				// ensure the file list is an empty collection (not nil)
-				metadata.Files = make([]pkg.DpkgFileRecord, 0)
-			}
+		results = append(results, pkgs...)
+	}
+	return results, nil
+}
 
-			// persist alterations
-			p.Metadata = metadata
+func addLicenses(resolver source.FileResolver, dbLocation source.Location, p *pkg.Package) {
+	// get license information from the copyright file
+	copyrightReader, copyrightLocation := fetchCopyrightContents(resolver, dbLocation, p)
 
-			// get license information from the copyright file
-			copyrightReader, copyrightLocation, err := fetchCopyrightContents(resolver, dbLocation, p)
-			if err != nil {
-				return nil, fmt.Errorf("unable to find dpkg copyright contents: %w", err)
-			}
+	if copyrightReader != nil {
+		// attach the licenses
+		p.Licenses = parseLicensesFromCopyright(copyrightReader)
+
+		// keep a record of the file where this was discovered
+		if copyrightLocation != nil {
+			p.Locations = append(p.Locations, *copyrightLocation)
+		}
+	}
+}
 
-			if copyrightReader != nil {
-				// attach the licenses
-				p.Licenses = parseLicensesFromCopyright(copyrightReader)
+func mergeFileListing(resolver source.FileResolver, dbLocation source.Location, p *pkg.Package) {
+	metadata := p.Metadata.(pkg.DpkgMetadata)
 
-				// keep a record of the file where this was discovered
-				if copyrightLocation != nil {
-					p.Locations = append(p.Locations, *copyrightLocation)
-				}
+	// get file listing (package files + additional config files)
+	files, infoLocations := getAdditionalFileListing(resolver, dbLocation, p)
+loopNewFiles:
+	for _, newFile := range files {
+		for _, existingFile := range metadata.Files {
+			if existingFile.Path == newFile.Path {
+				// skip adding this file since it already exists
+				continue loopNewFiles
```

Comment:
This looks like it isn't doing anything... I don't see what is the condition where it would do something else when the file doesn't already exist

Reply:
The key to this logic is the argument to `continue`.

There are two loops here; the outer one is a named loop (`loopNewFiles`) looping through the new files and an inner loop that is checking the existing files. The last line in the outer loop (`loopNewFiles`) will append the new file. However, the `continue loopNewFiles` statement occurs when the new path matches an existing path. The optional argument to `continue` here is key, as it is noting to continue the outer loop iteration and not the inner loop iteration.

Reply:
oh wow - this is something I'd not seen before, thanks for explaining it